### ISSUE TITLE
Feature request: Support for model-averaged GAMs

### ISSUE THREAD

**Issue by crsh** [rvlenth/emmeans#124]
I'm trying to use `emmeans` to compare estimated marginal means averaged over a set of GAMs. I'm using `MuMIn` to average models. However, as noted in the documentation, the support for averaged models is "dodgy, as it is not clear that all supported model classes will work." GAMs, currently, seem to be unsupported. When I use `emmeans()` or `ref_grid()` on the average object I get the following error:

~~~r
library("MuMIn")
library("mgcv")
library("emmeans")

set.seed(2) ## simulate some data... 
dat <- gamSim(1, n = 400, dist = "normal", scale = 2)
mod <- gam(y ~ s(x0) + s(x1) + s(x2) + s(x3), data = dat, na.action = "na.fail")

gam_dredged <- dredge(mod, rank = "AIC")

gam_averaged <- model.avg(gam_dredged, fit = TRUE)

ref_grid(
  gam_averaged
  , data = dat
)
~~~
~~~
Error in model.frame.default(formula, data = data, ...) : 
  invalid type (list) for variable 's(x0)'
~~~

I've tried to zero in on the problem but it seems I'm not getting anywhere. I tried to circumvent the issue by passing `trms` manually but get a different error:

~~~r
ref_grid(
  gam_averaged
  , data = dat
  , trms = delete.response(terms(y ~ x0 + x1 + x2 + x3))
)
~~~
~~~
Error in `[.terms`(tbl, , vars, drop = FALSE) : 
  unused arguments (vars, drop = FALSE)
~~~

From what I was able to make out, this error may be addressed by specifying `data = data` to `recover_data.averaging()` [here](https://github.com/rvlenth/emmeans/blob/6f1237d1f4fb04b2cc09f5984f32ecd5d0e81a08/R/multiple-models.R#L45). However, doing this leaves me with yet another error:

~~~
Error in match(x, y, 0L) : 'match' requires vector arguments 
~~~

In short, I've tried but failed to understand why the `averaging`-methods are failing for GAMs. Is there something I'm misunderstanding or not  seeing? Would you be willing to add support for averaged GAMs?

**Comment by rvlenth** [comment]
Sorry for delayed response; travel and family visits.

The problem is that `gam` objects with smooting components have a very complex structure, and if you are averaging several together, you need to account for separate spline bases for each model. Meanwhile, the design for `averaging` objects relies on setting up a common **X** matrix that works with all of the models, tracking which regression coefficients are zero in each model. 

So, as currently implemented, it is basically impossible to support averaged `gam` models in **emmeans** -- only one `gam` at a time is possible. Sorry. I'll think about whether there is a different design that could work efficiently.

**Comment by nahorp** [comment]
Piggybacking on this, I am trying to this with `glmer` objects and am coming across an error that is different from the above. I've checked the averaged model and, as per the vignette, it has the `modelList` attribute as well as the call containing the `data` argument. Are `glmer` objects also not supported or am I doing something stupid (please let it be the latter!)?

```
library("MuMIn")
library("emmeans")

##Creating data

set.seed(666)
x1 = rnorm(50)           
x2 = rnorm(50)
y = rbinom(50,1,0.4)
id = (1:50)
df = data.frame(y=y,x1=x1,x2=x2,id=id)

##Running model

model <- glmer(y ~ x1 + x2 + (1|id), data=df, family="binomial", na.action = "na.fail")
model_dredged <- dredge(model, rank = "AIC")
model_averaged <- model.avg(model_dredged, fit = TRUE, subset = delta < 1.5, data = df)
emmeans(model_averaged, "x1")
```
The error that is spit out is as follows,

```
Error : $ operator not defined for this S4 class
Error in ref_grid(object, ...) : 
Perhaps a 'data' or 'params' argument is needed
```

**Comment by rvlenth** [comment]
I am traveling right now, but did you try

    emmeans(model_averaged, "x1", `data = df)

Maybe that’s all it needs. But maybe not. I’ll check later when I can.

**Comment by nahorp** [comment]
I did try that (i.e., `emmeans(model_averaged, "x1", data = df)`). The error it spits out,

`Error: $ operator not defined for this S4 class`

That'd be great Russell, thank you. Enjoy the travels :)

**Comment by rvlenth** [comment]
There are different things I see happening. One is tyhis, when I load the MuMin package:
```r
> library("MuMIn")
Registered S3 method overwritten by 'MuMIn':
  method         from
  predict.merMod lme4
```
This is not good, because it means there are different predict methods for `MuMin` models depending on which package was loaded first. That may or may not make a difference, but it is outside of **emmeans**'s purview.

The remaining issues -- and these *are* in **emmeans** -- are

  * In this particular model, `data` must be specified. You can work around it by first creating a list of models using `model.list()` before calling `model.avg()`
  * The failure then occurs because `glmer()` creates an S4 object, and my code wants `object$family` and `object$df.residual`. Those don't work, and that's why you get the error. I have changed these to calls to the `family()` and `df.residual()` accessors, and now it works:
```r
> emmeans(model_averaged, "x1", data = df)
      x1 emmean    SE df lower.CL upper.CL
 -0.0648 -0.325 0.288 47   -0.903    0.254

Results are given on the logit (not the response) scale. 
Confidence level used: 0.95 
```
I'll push this patched version and reference this issue.

**Comment by rvlenth** [comment]
I'm closing this issue, as I think it is resolved as much as possible at this time. I do recognize that averages of `gam` models are still not supported. My long-term goal is to try to shift **emmeans** support to developers/maintainers of the packages that implement the models. That will help with cases where the structure of model objects changes with a package update.

**Comment by nahorp** [comment]
Apologies about the delay Russell, I've finally had some time to try it out and it works now, so thank you! :) 👍 Just one question, the functions (emmeans, joint_tests etc.) all utilise the 'full averaged' model coefficients. Is there an argument in those functions to use the 'conditional averaged' coefficients instead?

**Comment by rvlenth** [comment]
Here's the problem with that. The architecture of the `emmGrid` object, for better or worse, involves the fixed-effects coefficients `object@bhat`, their covariance matrix, `object@V`, and the linear functions we need to compute, `object@linfct`. Then for each row `x` of `linfct`, our EMM is computed as `t(x) %*% bhat`, and its estimated variance is `t(x) %*% V %*% x`. 

However, with the conditional averaged coefficients, the `V` matrix is not necessarily positive-definite; and that means that the estimated variance can go negative, which makes everything go tilt. I found this out the hard way in an early test with my support for **MuMIn**. 

I think in the context of model averaging, what is usually done is we compute the estimates and SEs from each model. We then average together the estimates, and somehow average together the SEs along with an adjustment for the between-model variance. For the full averaged model, that approach is consistent with **emmeans**'s architecture; but the conditional averaged model isn't consistent with my internal design.

**Comment by nahorp** [comment]
I see, so the onus would be on the creators of **MuMIn** to create the conditional averaged model that is consistent with **emmeans**'s internal architecture?

Also, when you said 

> We then average together the estimates, and somehow average together the SEs

In link with the the 'Timing is everything' and 'The model is our best guide' subsections of the 'Transformations and link functions in emmeans' vignette, the averaging is all done on the link scale, right?

**Comment by rvlenth** [comment]
I’m on a vacation,so this’ll be brief.

The internal architecture of emmeans is that everything is linear. Complete averages are linear, and conditional averages are not. I’m not sure it’s even possible to support conditional averaging

Russ

Sent from my iPad

On Aug 28, 2019, at 6:01 PM, Rohan Puri <notifications@github.com<mailto:notifications@github.com>> wrote:


I see, so the onus would be on the creators of MuMIn to create the conditional averaged model that is consistent with emmeans's internal architecture?

Also, when you said

We then average together the estimates, and somehow average together the SEs

In link with the the 'Timing is everything' and 'The model is our best guide' subsections of the 'Transformations and link functions in emmeans' vignette, the averaging is all done on the link scale, right?
